planner/memo/expr_iterator.go

```diff
@@ -161,6 +169,9 @@ func newExprIterFromGroupExpr(expr *GroupExpr, p *Pattern) *ExprIter {
 
 // newExprIterFromGroup creates the iterator on the Group.
 func newExprIterFromGroup(g *Group, p *Pattern) *ExprIter {
+	if p.Operand == OperandAny {
+		return &ExprIter{Group: g, Operand: OperandAny, matched: true}
```

Comment:
From my understanding, `Any` should match any operator or operator sub-tree, otherwise, for a single rule such as `PushSelDownTableGather`, we have to enumerate too many patterns, such as `Selection -> TableGather -> DataSource`, `Selection -> TableGather -> Selection -> DataSource`, `Selection -> TableGather -> Selection -> Selection -> DataSource` ...

Reply:
But for sub tree case, we will need to check it in the operator itself, that is another pattern matching code. For the above case, I think `Selection -> TableGather -> Any` and one rule to merge selection is enough.

Reply:
Sorry I did't get your point, could you please elaborate more on this?

Reply:
If `Any` can match any subtree, then it loss the ability to specifiy patterns like `Any -> XXX`, but if this kind of patterns is common, the the subtree matching has to be done within operator itself, it is yet another pattern matching code.

Reply:
Well, I understand what @lamxTyler means. IMO, when we uses `X -> Any`, we only want to make sure there is a node below `X`, so that we can safely uses `X.Children[0]`. As for `Any -> XXX`, do we really need such pattern? If so, it seems we also have to split `Any` into `UnaryAny` and `BinaryAny`.

If we want to do more pattern matching before `OnTransform`, we may need another method to represent the `Operand` such as bitmap.

Reply:
Hmm, @lamxTyler if `Any` would just match a single operator, then for the rule here, the pattern should be `Selection -> TableGather -> Any`, `Selection -> TableGather -> Any -> Any`, etc. Also, as @francis0407 mentioned, `Any` need to be split into `UnaryAny` and `BinaryAny` and `MultiAny`(for UnionAll)?

Reply:
For patterns like `X -> Any -> Y`,  we can still insist `Any` to match any operator or subtree, and check the `-> Y` constraint in `Match` interface of the rule. @lamxTyler how about this workaround?

Reply:
OK, and we can refactor code later if we found best practice that matching our usage.

Reply:
From the beginning, `Any` is used to represent a node in the plan tree.

For the pattern `X -> Any`, if `Any` is some node that doesn't have any child, it represents a single node. But if it has some children, it represents a sub-tree.

For the pattern `Any -> X`, `Any` just means we don't care what the node type is. But it has to have a child which is `X`. This explanation also suit for the scenario of `X -> Any -> Y`.  

And I think we don't need to have `UnaryAny`, `BinaryAny` or `MultiAny` under this definition.